src/transactions/test/SorobanTxTestUtils.cpp

```diff
@@ -571,6 +626,500 @@ ContractInvocationTest::createUploadWasmTx(TestAccount& source,
+        ltx.commit();
+    }
+
+    mContractKeys.emplace_back(contractExecutableKey);
+    mContractID = contractID;
+}
+
+void
+AssetContractInvocationTest::transfer(TestAccount& fromAcc,
+                                      SCAddress const& toAddr, int64_t amount,
+                                      bool expectSuccess)
+{
+    SCVal toVal(SCV_ADDRESS);
+    toVal.address() = toAddr;
+
+    SCAddress fromAddr(SC_ADDRESS_TYPE_ACCOUNT);
+    fromAddr.accountId() = fromAcc.getPublicKey();
+    SCVal fromVal(SCV_ADDRESS);
+    fromVal.address() = fromAddr;
+
+    LedgerKey fromBalanceKey;
+    if (mAsset.type() == ASSET_TYPE_NATIVE)
+    {
+        fromBalanceKey.type(ACCOUNT);
+        fromBalanceKey.account().accountID = fromAcc.getPublicKey();
+    }
+    else
+    {
+        fromBalanceKey.type(TRUSTLINE);
+        fromBalanceKey.trustLine().accountID = fromAcc.getPublicKey();
+        fromBalanceKey.trustLine().asset = assetToTrustLineAsset(mAsset);
+    }
+
+    LedgerKey toBalanceKey;
+    if (toAddr.type() == SC_ADDRESS_TYPE_ACCOUNT)
+    {
+        if (mAsset.type() == ASSET_TYPE_NATIVE)
+        {
+            toBalanceKey.type(ACCOUNT);
+            toBalanceKey.account().accountID = toAddr.accountId();
+        }
+        else
+        {
+            toBalanceKey.type(TRUSTLINE);
+            toBalanceKey.trustLine().accountID = toAddr.accountId();
+            toBalanceKey.trustLine().asset = assetToTrustLineAsset(mAsset);
+        }
+    }
+    else
+    {
+        toBalanceKey.type(CONTRACT_DATA);
+        toBalanceKey.contractData().contract = mContractID;
+
+        SCVec balance = {makeSymbolSCVal("Balance"), toVal};
+        SCVal balanceKey(SCValType::SCV_VEC);
+        balanceKey.vec().activate() = balance;
+        toBalanceKey.contractData().key = balanceKey;
+        toBalanceKey.contractData().durability =
+            ContractDataDurability::PERSISTENT;
+    }
+
+    auto fn = makeSymbol("transfer");
+    Operation transfer;
+    transfer.body.type(INVOKE_HOST_FUNCTION);
+    auto& ihf = transfer.body.invokeHostFunctionOp().hostFunction;
+    ihf.type(HOST_FUNCTION_TYPE_INVOKE_CONTRACT);
+    ihf.invokeContract().contractAddress = mContractID;
+    ihf.invokeContract().functionName = fn;
+    ihf.invokeContract().args = {fromVal, toVal, makeI128(amount)};
+
+    transfer.body.invokeHostFunctionOp().auth = {
+        getInvocationAuth(ihf.invokeContract())};
+
+    SorobanResources resources;
+    resources.instructions = 2'000'000;
+    resources.readBytes = 2000;
+    resources.writeBytes = 1072;
+
+    resources.footprint.readOnly = mContractKeys;
+
+    bool fromIsIssuer = false;
+    bool toIsIssuer = false;
+    if (mAsset.type() != ASSET_TYPE_NATIVE)
+    {
+        LedgerKey issuerLedgerKey(ACCOUNT);
+        issuerLedgerKey.account().accountID = getIssuer(mAsset);
+        resources.footprint.readOnly.emplace_back(issuerLedgerKey);
+
+        if (!(getIssuer(mAsset) == fromAcc.getPublicKey()))
```

Comment:
The inequality operator isn't defined, which is why we have to use the equality operator here.

Reply:
Sorry, I meant that we should change the if check to `getIssuer(mAsset) == fromAcc.getPublicKey()` and swap the body of the if and else to avoid the not ==. At least to me it seems clearer, lots of parenthesis and operators take me a bit to decipher.

Reply:
Ah yeah fixed